pangeo/jobqueue.py

```diff
@@ -0,0 +1,126 @@
+    """
+    def __init__(self):
+        logger.error("JobQueue class should not be called directly")
+
+    def job_script(self):
+        self.n += 1
+        return self._template % toolz.merge(self.config, {'n': self.n})
+
+    @contextmanager
+    def job_file(self):
+        """ Write job submission script to temporary file """
+        with tmpfile(extension='sh') as fn:
+            with open(fn, 'w') as f:
+                f.write(self.job_script())
+            yield fn
+
+    def start_workers(self, n=1):
+        """ Start workers and point them to our local scheduler """
+        outs = []
+        workers = []
+        for _ in range(n):
+            with self.job_file() as fn:
+                out = self._call([self._submitcmd, fn])
+                job = out.decode().split('.')[0]
+                self.jobs[self.n] = job
+                workers.append(self.n)
+        return workers
+
+    @property
+    def scheduler(self):
+        return self.cluster.scheduler
+
+    @property
+    def scheduler_address(self):
+        return self.cluster.scheduler_address
+
+    def _calls(self, cmds):
+        """ Call a command using subprocess.communicate
+
+        This centralzies calls out to the command line, providing consistent
+        outputs, logging, and an opportunity to go asynchronous in the future
+
+        Parameters
+        ----------
+        cmd: List(List(str))
+            A list of commands, each of which is a list of strings to hand to
+            subprocess.communicate
+
+        Examples
+        --------
+        >>> self._calls([['ls'], ['ls', '/foo']])
+
+        Returns
+        -------
+        The stdout result as a string
+        Also logs any stderr information
+        """
+        logger.debug("Submitting the following calls to command line")
+        for cmd in cmds:
+            logger.debug(' '.join(cmd))
+        procs = [subprocess.Popen(cmd,
+                                  stdout=subprocess.PIPE,
+                                  stderr=subprocess.PIPE)
+                 for cmd in cmds]
+
+        result = []
+        for proc in procs:
+            out, err = proc.communicate()
+            if err:
+                logger.error(err.decode())
+            result.append(out)
+        return result
+
+    def _call(self, cmd):
+        """ Singular version of _calls """
+        return self._calls([cmd])[0]
+
+    def stop_workers(self, workers):
+        if not workers:
+            return
+        workers = list(map(int, workers))
+        jobs = [self.jobs[w] for w in workers]
+        self._call([self._cancelcmd] + list(jobs))
+        for w in workers:
+            with ignoring(KeyError):
+                del self.jobs[w]
+
+    def scale_up(self, n, **kwargs):
+        return self.start_workers(n - len(self.jobs))
+
+    def scale_down(self, workers):
+        if isinstance(workers, dict): # https://github.com/dask/distributed/pull/1659
```

Comment:
I don't think this comment is useful is needed anymore. This feature is not in distributed@master.